### ISSUE TITLE
HHVM main issues from a user's prespective

### ISSUE THREAD

**Issue by fesarlis** [facebook/hhvm#7750]
Hello all.
I am not sure that this is the place to post my experience or this is a contributor-only forum. Please forgive me in advance.

So I decided to switch to HHVM for my WP sites, as a result of my disappointment from PHP in terms of performance. The first experience is rather positive, but I am facing some important issues and it seems to me there are not enough resources that can help, as most of them are rather old.

The main problem is code compiling. It throttles both my CPUs to 100% each time a page is requested, and  the strange thing is that it also happens even when a page was served before (same URL). I have not been able to solve this by tuning anything. I also could not find info on how to pre-analyze my websites (about 20 Apache virtual hosts). I have also tried preauthoritative mode and all the sites were broken. I guess it was my mistake but don't know how to fix it. So:

1. Is pre-analyzing (hhvm --hphp ...) and preauthoritative mode the same thing?
2. How can I pre-analyze all websites without making the server unresponsive? It is a production server so this is not an option. How will I use the generated file? Is it the same as the auto-generated hhvm.hhbc?
3. Setting hhvm.jit to false seems to solve the above issues. However I am not sure if doing so will result in HHVM having the same speed as PHP 5.6.30 I was using previously.
4. Is HHVM compression needed when there is web server level compression (apache mod_deflate)?

I would appreciate some guidance.
Thank you

**Comment by mofarrell** [comment]
If you are using apache, make sure you are using FastCGI, and not just CGI (http://hhvm.com/blog/1817/fastercgi-with-hhvm).  CGI starts a process for each request, and terminates it at the end.  This means HHVM would lose all the jitted code from the request.  FastCGI will have HHVM service requests without starting and stopping it.  This lets a single HHVM process service many requests, and build up a JITed translation cache.  As a rule of thumb, there should only ever be a single HHVM process running.  HHVM uses threads to service requests.  This is unlike PHP which uses processes.

In an ideal scenario, you eliminate Apache from the system altogether, and use HHVM's built in webserver Proxygen.  Proxygen allows vhost configuration like Apache, and should lead to better performance.  Here are the configuration options for vhost https://docs.hhvm.com/hhvm/configuration/INI-settings#server-mode__virtual-host-format.

At the same time this makes me wonder how you are running wordpress.  Are you using wordpress multisite?  Or do you have a bunch of copies of wordpress?  HHVM currently does not collapse identical files.  This means it will JIT each copy of wordpress rather than realizing they represent the same code.  Running using multisite should lead to better performance.  That being said there should be nothing wrong with having multiple copies.

You mentioned repo-authoritative mode (what you were calling preauthoritative).  This mode gives an additional boost of performance ~5-10%, but it is more important to make sure the JIT is functioning properly before playing with repo-authoritative mode.  Repo-authoritative mode does not do any ahead of time JITing (that would make it normal compilation); instead it does a global optimization of HHBC (hiphop bytecode).  When you run in repo-authoritative mode the PHP code being run cannot be modified, so HHBBC (hiphop byte code to byte code compiler) can do global optimizations rather than just file local ones.

**Comment by fesarlis** [comment]
Thank you for your immediate and detailed response.

Eliminating Apache is not an option at the moment, due to company policy issues. Please also note that several WP sites don't work with HHVM so PHP has to remain active (FYI sites that using WP Types family of plugins). Of course we use HHVM  via FastCGI. We tried using it via TCP/IP, and now we are using the socket.

In any case, no, we don't use multisite.What you mention sounds quite logical, but I have to make clear that I am talking about the same WP file on the SAME website being re-compiled.

I will not deal with repo-authoritative mode at the moment, to keep things simple. Yet my initial questions still stand, so let me express them in a different way:

1. Multi threads are used as far as I can tell. Did not change any variables about this (thread count etc).

2. As I see it, code compiling should occur in background. Or at least, could it be adjusted not to use 100% of the CPU, since it also makes other sites with compiled files unresponsive.

3. Please clarify what happens if jit.enabled=false. Although I dont like it, I really must use it or disable HHVM, as I have had many complaints from users about delays. 

4. The PERFECT scenario for me (and I guess for many users) would be:
   a) Disable HHVM JIT
   b) manually walk all web server directories, find php files and compile them WITHOUT using 100% of the CPU.
   c) Stop HHVM
   d) replace hhvm.hhbc with the file generated in (b)
   e) Start HHVM and never notice other delays until I update WP.

   Can this be done?

Thanks again

Thanks

**Comment by mofarrell** [comment]
So hhvm.hhbc is not JITed code.  It is similar to PHP opcache. It is a cache of opcodes that do high level functionality (The spec is [here](https://github.com/facebook/hhvm/blob/master/hphp/doc/bytecode.specification)), not machine code.  There is no way to JIT ahead of time (plus it wouldn't be a JIT then).  We can't do this because of dynamic typing.  Additionally runtime JITing allows us to profile code, and emit translations that take into account usage of different code paths.  I mentioned options for controlling the speed of the JIT on this issue https://github.com/facebook/hhvm/issues/7747

Can you confirm that your machine only has a single HHVM instance running?  Or is it an instance per wordpress site?  Running multiple copies of HHVM means it will over consume resources because the runtime option defaults assume it is running alone on a 16+ core machine
How many cores does the machine have?  What is your process for warming up the site?  [This covers warmup](https://github.com/facebook/hhvm/wiki/performance-tuning#warm-up-your-server---10).  After startup hitting your most popular endpoints will trigger the JIT so future actual requests will hit JITed code.

If you want better performance and throughput don't set the JIT to false.  The JIT is what gives HHVM an advantage over PHP.

Also apparently the hashes identifying units are based only on the filename and contents not the entire path.  This means that the units should be collapsed unlike I was saying originally.

**Comment by fesarlis** [comment]
My machine has only two cores, sadly.
Actually, I did not perform any warm-up at all. I find the process rather confusing (although I did not spend much time reading about it), but basically I thought that after a while warm-up will be done automatically by website visitors (although it would cause  some delay in first requests). However, and due to the fact that they are not very busy sites (200-500 hits /day) this is taking more than I expected. To put it simply: Is warm-up absolutely necessary or it the server will stabilise after a while? (in my case, some days)

As for the unit identification. Based on what you are saying, this could be a problem in my case, because not all sites have the same WP version. So a file.php in one site could be different than file.php in other site. Is HHVM equiped to handle this or this will result in a loop compiling file.php when both websites are visited frequently?

Also, I am not sure where JIT code is stored, and if it was created at all.

**Comment by mofarrell** [comment]
JITed code will be stored in some heap space HHVM allocates.  The unit problem shouldn't be a major concern.  The warm-up process should happen naturally over time.  Explicitly warming the server up might lead to users having a better experience as they won't hit the interpreter.

I believe you will have better performance if you dial back `hhvm.jit_threads`.  It defaults to 4, which is your entire machine.  This should control the number of threads used by the concurrent JIT http://hhvm.com/blog/2017/03/23/concurrent-jit-compilation.html.  Its worth noting that the JIT process will take time.  First profiling translations are created, then after a given translation is hit more than `hhvm.jit_pgo_threshold` times it will be retranslated as an optimized translation.  It might be worth tuning that downwards if you don't think there will be a lot of hits.  At the same time it shouldn't make that big a difference.

**Comment by fesarlis** [comment]
I understand. Currently I am using HHHV 3.15.3 under CentOS 7.3. I tried to compile it successfully, but I ran into some problems when tried to run it, so I dropped it for now. I notice that the variables you mention are not in the documentation. Are they supported in 3.15.3 or do I have to update? I tried setting both `hhvm.jit_threads` and `hhvm.jit_worker_threads`  to 1 to start with (as you suggested in the other post), but I did not notice any difference (CPU still at 95-100%).

**Comment by mofarrell** [comment]
@fesarlis It would be helpful if you could provide an example set of configuration files, and code that produces the behavior you are mentioning.  Without an isolated case, I am just guessing as to what is happening.

Also the `jit_threads` option will be in 3.15, but the `jit_worker_threads` will not be.

**Comment by fesarlis** [comment]
Okay, I now have some new situation. I have upgraded to a newer and stronger server, mainly comprising of an i7-3770 and 16GB of RAM. So compilation speed is not so much a problem now. I will just need some guidance to optimize HHVM's performance for it, and ideally use all cores for all PHP intensive procedures.

I would appreciate your help. In any case, I am so grateful that you've spent so much of your valuable time to deal with my case. I really appreciate it.

**Comment by fesarlis** [comment]
Please also let me include my server.ini for reference:

; php options

pid = /var/run/hhvm/pid

`
session.save_handler = files
session.save_path = /tmp
session.gc_maxlifetime = 1440
post_max_size = 100M
upload_max_file_size = 100M
expose_php=false
max_execution_time=300
max_file_uploads = 50
error_reporting = E_ALL & ~E_NOTICE & ~E_STRICT & ~E_DEPRECATED
date.timezone = "Europe/Athens"

; hhvm specific

hhvm.server.port = 9001
hhvm.server.type = fastcgi
hhvm.server.default_document = index.php
hhvm.repo.central.path = /opt/hhvm/hhvm.hhbc
hhvm.server.file_socket = /tmp/hhvm.sock
hhvm.server.enable_keep_alive = true
hhvm.server.apc.enable_apc = true
hhvm.server.request_timeout_seconds = 120
hhvm.server.expose_hphp = false

hhvm.jit = true
hhvm.jit_threads = 8

hhvm.log.level = Notice
hhvm.log.always_log_unhandled_exceptions = true
hhvm.log.runtime_error_reporting_level = 8191
hhvm.log.use_log_file = true
hhvm.log.use_syslog = false
hhvm.log.file = /var/log/hhvm/error.log
hhvm.log.header = true
hhvm.log.native_stack_trace = true

hhvm.mysql.socket = /tmp/mysql.sock
hhvm.mysql.typed_results = false
hhvm.mysql.slow_query_threshold = 5000
; php options

pid = /var/run/hhvm/pid

session.save_handler = files
session.save_path = /tmp
session.gc_maxlifetime = 1440
post_max_size = 100M
upload_max_file_size = 100M
expose_php=false
max_execution_time=300
max_file_uploads = 50
error_reporting = E_ALL & ~E_NOTICE & ~E_STRICT & ~E_DEPRECATED
date.timezone = "Europe/Athens"

; hhvm specific

hhvm.server.port = 9001
hhvm.server.type = fastcgi
hhvm.server.default_document = index.php
hhvm.repo.central.path = /opt/hhvm/hhvm.hhbc
hhvm.server.file_socket = /tmp/hhvm.sock
hhvm.server.enable_keep_alive = true
hhvm.server.apc.enable_apc = true
hhvm.server.request_timeout_seconds = 120
hhvm.server.expose_hphp = false

hhvm.jit = true
hhvm.jit_threads = 8

hhvm.log.level = Notice
hhvm.log.always_log_unhandled_exceptions = true
hhvm.log.runtime_error_reporting_level = 8191
hhvm.log.use_log_file = true
hhvm.log.use_syslog = false
hhvm.log.file = /var/log/hhvm/error.log
hhvm.log.header = true
hhvm.log.native_stack_trace = true

hhvm.mysql.socket = /tmp/mysql.sock
hhvm.mysql.typed_results = false
hhvm.mysql.slow_query_threshold = 5000
`

**Comment by mofarrell** [comment]
Your perf settings look fine.  I wouldn't increase the `jit_threads` setting.  The default `4` may even be a little high.  `8` will be all the computing resources on your CPU.  Most servers are running Xeons with high core counts, and some settings assume this kind of setup.

If you are running PHP (like wordpress), and therefore not using any async features, you may want to set `hhvm.server.thread_count`.  It defaults to `2 x CPU cores`.  This is a little low when work loads are seriously IO bound (on database accesses).  Something in the range of two to four times the number of CPU cores would be a good setting.

I am going to close this, as it doesn't appear to be an issue with HHVM.  You can continue commenting here, and I will reopen the issue if something actionable comes up.

**Comment by fesarlis** [comment]
Thanks again.

**Comment by mofarrell** [comment]
Daemon mode is the same as server mode, except that it shells out a new process, and daemonizes it (makes it won't die when you log out).

Wordpress plugins should work.  However, there are known compatibility issues with respect to PHP 5.

**Comment by fesarlis** [comment]
Sorry for reverting back but I have noticed that when JIT compiles a specific resource it only uses one core. I have tried to experiment with the settings you mentioned but the behaviour remains the same. 

It will utilize more cores only when it is required to  use more threads (eg it needs to compile more pages at the same time). But at any given point, one thread uses one core.

Is this normal?

**Comment by mofarrell** [comment]
Yes.  The concurrent JIT allows multiple requests to JIT at the same time, but currently does not allow parallel JITing of a single request.  This is roughly because: In order to know when a tracelet ends and the next one starts we must JIT it. At the start of each tracelet we also want to know the types of variables used in the tracelet, and because PHP is dynamically typed we don't know them unless there is a request running.  Basically we have to know what the request does in order to JIT the code, and the easy way to do that is to have the JIT run as a part of the request.

That being said the retranslation of the original profiling translations can be done in parallel and out of band from the request threads.  This is default at least in master.

**Comment by fesarlis** [comment]
I see. And, is this normal for JIT to recompile a page that has already been requested by the same client? This happens sometimes.

**Comment by mofarrell** [comment]
That shouldn't happen if there are existing translations for the types found at runtime.

**Comment by fesarlis** [comment]
Then perhaps something is happening from an external factor. I have noticed there are some perf-* files under /tmp. So could this mean that JITed code does not survive a reboot? If this is not the code, it would be really helpful to know where it is stored.

**Comment by mofarrell** [comment]
It doesn't survive a reboot.  It is JITed when HHVM starts, and dies when HHVM dies.  It is not recorded anywhere on the filesystem, it is stored in memory.  The perf-map files you mention give information about the symbol locations in the translation cache (still stored in memory) for debugging and profiling purposes.

**Comment by fesarlis** [comment]
Then clearly I must be doing something wrong, or do not understand the essence of HHVM. Every time a page that is not JITed is visited, the visitor is waiting a minimum of 20 seconds until the page is shown. HHVM crashes frequently (in my case) so is there a chance that HHVM is useless for someone like me and is meant to be used only in datacenter-level infrastructures? Taking 20-40 seconds to compile a simple WP page (a number of PHP files involved I guess) is kind of weird for an i7 class processor.

**Comment by mofarrell** [comment]
HHVM itself should not crash.  That is a sign of a pretty serious bug.  Fataling a request is more acceptable although it should not happen.

The very first loaded page may take 20-40 seconds, but you shouldn't have to restart HHVM for extended periods of time making this a one time cost.  Also each endpoint should share a lot of code, and be less slow than the very first request.  You can also warmup the server ahead of time making these slow request happen before users see the site (see https://github.com/facebook/hhvm/issues/5312 for more info).
If you can provide more information about the crashes you are seeing (like the stack traces), open a new issue, and I will see what I can do.  It would be best if you were running the latest version of HHVM, 3.18.  You may be encountering an issue that has already been fixed.
